src/unread/__tests__/unreadModel-test.js

```diff
@@ -63,6 +67,95 @@ describe('stream substate', () => {
+    const mkAction = args => {
+      const {
+        message_id,
+        message_ids = [message_id],
+        stream_id = 123,
+        new_stream_id = undefined,
+        orig_subject = 'foo',
+        subject = 'foo',
+      } = args;
+      return {
+        id: 1,
+        type: EVENT_UPDATE_MESSAGE,
+        edit_timestamp: 10000,
+        message_id,
+        message_ids,
+        stream_id,
+        new_stream_id,
+        orig_content: '',
+        orig_subject,
+        orig_rendered_content: '',
+        prev_rendered_content_version: 0,
+        rendered_content: '',
+        subject_links: [],
+        subject,
+        user_id: eg.selfUser.user_id,
+      };
+    };
+
+    const baseState = (() => {
+      const streamAction = args => eg.mkActionEventNewMessage(eg.streamMessage(args));
+      const r = (state, action) => reducer(state, action, eg.plusReduxState);
+      let state = initialState;
+      state = r(state, streamAction({ stream_id: 123, subject: 'foo', id: 1 }));
+      state = r(state, streamAction({ stream_id: 123, subject: 'foo', id: 2 }));
+      state = r(state, streamAction({ stream_id: 123, subject: 'foo', id: 3 }));
+      state = r(state, streamAction({ stream_id: 123, subject: 'foo', id: 4 }));
+      state = r(state, streamAction({ stream_id: 123, subject: 'foo', id: 5 }));
+      state = r(state, streamAction({ stream_id: 456, subject: 'zzz', id: 6 }));
+      state = r(state, streamAction({ stream_id: 456, subject: 'zzz', id: 7 }));
+      return state;
+    })();
+
+    test('if topic not updated, return original state', () => {
+      const state = reducer(baseState, mkAction({ message_id: 5 }), eg.plusReduxState);
+      expect(state.streams).toBe(baseState.streams);
+    });
+
+    test('if topic updated, but no unreads, return original state', () => {
+      const state = reducer(baseState, mkAction({ message_id: 100 }), eg.plusReduxState);
+      expect(state.streams).toBe(baseState.streams);
+    });
+
+    test('if topic updated, move unreads', () => {
+      const state = reducer(
+        baseState,
+        mkAction({
+          message_id: 3,
+          message_ids: [3, 4, 5],
+          orig_subject: 'foo',
+          subject: 'bar',
+        }),
+        eg.plusReduxState,
+      );
+      // prettier-ignore
+      expect(summary(state)).toEqual(Immutable.Map([
+        [123, Immutable.Map([['foo', [1, 2]], ['bar', [3, 4, 5]]])],
+        [456, Immutable.Map([['zzz', [6, 7]]])],
+      ]));
+    });
+
+    test('if stream updated, move unreads', () => {
+      const state = reducer(
+        baseState,
+        mkAction({
+          message_id: 3,
+          message_ids: [3, 4, 5],
+          new_stream_id: 456,
+        }),
+        eg.plusReduxState,
+      );
+      // prettier-ignore
+      expect(summary(state)).toEqual(Immutable.Map([
+        [123, Immutable.Map([['foo', [1, 2]]])],
+        [456, Immutable.Map([['zzz', [6, 7]], ['foo', [3, 4, 5]]])],
```

Comment:
Let's also have a test case where they get moved to an existing topic (like 456 > 'zzz'), showing that the list gets properly combined with the existing one.